### ISSUE TITLE
[BUG] Enable key hash range | When Enable key hash range is set to true only one reader is created with a partial topic range on one source, this causes a Flink application to miss messages and fail.

### ISSUE THREAD

**Issue by avimas** [streamnative/pulsar-flink#230]
My setup is as follows:

1. Flink application with source parallelism of 5 
2. Pulsar Flink connector 2.7.1.1 with key hash range enabled 
3. One pulsar topic without partitions
 
When the application starts, only one reader with a partial range is created for the topic, and this happens only on one subtask/source, this means that messages that are not within the hash range can't be received , and Flink parallelism is still bounded to number of partitions.

The expected behavior is that each of the 5 sources would create a reader with a different hash range, and all the subtasks will consume the whole range in parallel.

As you can see in the logs below only source 4 created a reader but the range is partial

14:13:11.648 [Source: EventsInputStream (5/5)] INFO  [] o.a.f.s.c.pulsar.FlinkPulsarSource [] - Source 4 initially has no topics to read from.
14:13:11.648 [Source: EventsInputStream (1/5)] INFO  [] o.a.f.s.c.pulsar.FlinkPulsarSource [] - Source 0 initially has no topics to read from.
14:13:11.648 [Source: EventsInputStream (2/5)] INFO  [] o.a.f.s.c.pulsar.FlinkPulsarSource [] - Source 1 initially has no topics to read from.
14:13:11.648 [Source: AssetsStream (2/2)] INFO  [] o.a.f.s.c.pulsar.FlinkPulsarSource [] - Source 1 initially has no topics to read from.
14:13:11.648 [Source: EventsInputStream (3/5)] INFO  [] o.a.f.s.c.pulsar.FlinkPulsarSource [] - Source 2 initially has no topics to read from.
14:13:11.648 **[Source: EventsInputStream (4/5)]** INFO  [] o.a.f.s.c.pulsar.FlinkPulsarSource [] - Source 3 will start reading 1 topics from initialized positions: {TopicRange{topic=persistent://public/default/inputevents, key-range=SerializableRange{range=[39322, 52428]}}=9223372036854775807:9223372036854775807:-1}

**Comment by avimas** [comment]
After changing SourceSinkUtils belongs to method to:

public static boolean belongsTo(TopicRange topicRange, int numParallelSubtasks, int index) {
        String topic = topicRange.getTopic();
      
        **if (!(topicRange.isFullRange()) && rangesAreEqual(topicRange.getRange().getPulsarRange(), distributeRange(numParallelSubtasks, index))) {
            return true;
        }**

        if (topic.contains(PulsarOptions.PARTITION_SUFFIX)) {
            int pos = topic.lastIndexOf(PulsarOptions.PARTITION_SUFFIX);
            String topicPrefix = topic.substring(0, pos);
            String topicPartitionIndex = topic.substring(pos + PulsarOptions.PARTITION_SUFFIX.length());

            if (topicPartitionIndex.matches("0|[1-9]\\d*")) {
                int startIndex = (topicPrefix.hashCode() * 31 & Integer.MAX_VALUE) % numParallelSubtasks;
                return (startIndex + Integer.valueOf(topicPartitionIndex)) % numParallelSubtasks == index;
            }
        }
        return (topic.hashCode() * 31 & Integer.MAX_VALUE) % numParallelSubtasks == index;

    }

 private static boolean rangesAreEqual(Range pulsarRange, Range distributeRange) {
        return pulsarRange.getStart() == distributeRange.getStart() && pulsarRange.getEnd() == distributeRange.getEnd();
    }

the ranges are assigned correctly, but it raised another issue with the client creation.
After manually providing a Pulsar client everything worked.

I think that FlinkPulsarSource should have the option to get pulsarClient as a parameter

**Comment by avimas** [comment]
After I implemented the logic

**Comment by jianyun8023** [comment]
> Thanks a lot for looking at this.
> 
> After I implemented the logic above we had an issue that multiple clients were created(as the number of readers) instead of just one client, it looks like an issue with the cached client logic, after providing only one client everything worked.
> 
> I think that it would be better if the client of the library can provide a pulsar client as a dependency, instead of coupling the client to the library.
> 
> Then we encountered issue #248 causing the state to not being restored after parallelism change.
> The algorithm suggested by Arye Mazuz at #248 seems to handle this perfectly, please consider integrating it.

Relying on pulsar-client, we plan to provide a shade-free version in the near future.

I will take your suggestion into serious consideration.

**Comment by gaozhangmin** [comment]
any updates ?

**Comment by syhily** [comment]
I think this issue should be resolved by the latest connector in the Flink repository. Can you have a try on it?
https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/

**Comment by nlu90** [comment]
- [X] Merge into Master
- [ ] Cherry-pick into release-1.13
- [ ] Cherry-pick into release-1.14
